crates/subspace-farmer/src/commitments/commitment_databases.rs

```diff
@@ -95,62 +103,125 @@ impl CommitmentDatabases {
+
+        // Open the first one
+        if let Some(first_salt) = first_salt {
+            let db = DB::open(
+                &Options::default(),
+                base_directory.join(hex::encode(first_salt)),
+            )
+            .map_err(CommitmentError::CommitmentDb)?;
+            commitment_databases
+                .first_db
+                .write()
+                .replace(Arc::new(DbEntry {
+                    salt: *first_salt,
                     db: Mutex::new(Some(Arc::new(db))),
-                }),
-            );
+                }));
         }
+        // Open the second one
+        if let Some(second_salt) = second_salt {
+            let db = DB::open(
+                &Options::default(),
+                base_directory.join(hex::encode(second_salt)),
+            )
+            .map_err(CommitmentError::CommitmentDb)?;
+            commitment_databases
+                .second_db
+                .write()
+                .replace(Arc::new(DbEntry {
+                    salt: *second_salt,
+                    db: Mutex::new(Some(Arc::new(db))),
+                }));
+        }
+        drop(read_lock);
 
         Ok::<_, CommitmentError>(commitment_databases)
     }
 
     /// Get salts for all current database entries
     pub(super) fn get_salts(&self) -> Vec<Salt> {
-        self.databases
+        self.metadata
+            .read()
+            .metadata_cache
             .iter()
-            .map(|(salt, _db_entry)| *salt)
+            .map(|(salt, _commitment_status)| *salt)
             .collect()
     }
 
-    pub(super) fn get_db_entry(&self, salt: &Salt) -> Option<&Arc<DbEntry>> {
-        self.databases.peek(salt)
+    pub(super) fn get_db_entry(&self, salt: &Salt) -> Option<Arc<DbEntry>> {
+        return match salt {
+            _ if &self.first_db.read().as_ref().unwrap().salt == salt => {
+                self.first_db.read().clone()
+            }
+            _ if &self.second_db.read().as_ref().unwrap().salt == salt => {
+                self.second_db.read().clone()
+            }
+            _ => None,
+        };
     }
 
     /// Returns `Ok(None)` if entry for this salt already exists.
     pub(super) fn create_db_entry(
-        &mut self,
+        &self,
         salt: Salt,
+        current_salt: Salt,
     ) -> Result<Option<CreateDbEntryResult>, CommitmentError> {
-        if self.databases.contains(&salt) {
-            return Ok(None);
-        }
+        match salt {
+            _ if self.first_db.read().as_ref().is_some()
+                && self.first_db.read().as_ref().unwrap().salt == salt =>
+            {
+                return Ok(None)
+            }
+            _ if self.second_db.read().as_ref().is_some()
+                && self.second_db.read().as_ref().unwrap().salt == salt =>
+            {
+                return Ok(None)
+            }
+            _ => {} // don't do anything yet if salt does not exist
+        };
```

Comment:
What is this mess? `match` here does nothing, just like below.

Reply:
Yes, this part changed so much that, I did not even realize the `match` became useless along the process

Reply:
However, when I refactor them with `if else`, it is becoming less readable... And I believe in the end it is compiling to the same machine code. Do you prefer `if else`, or have another suggestion?

Reply:
How is `match { _ }` on top of existing `if` statements makes it more readable?

Reply:
Because it becomes obvious that only 1 of the paths will be taken. 4 `if else` chained together does not look imho.
```rust
if self.first_db.read().as_ref().is_none() {
            old_db_entry = self.first_db.write().replace(Arc::clone(&db_entry));
        } else if self.second_db.read().as_ref().is_none() {
            old_db_entry = self.second_db.write().replace(Arc::clone(&db_entry));
        } else if self.first_db.read().as_ref().is_some()
            && self.first_db.read().as_ref().unwrap().salt != current_salt
        {
            old_db_entry = self.first_db.write().replace(Arc::clone(&db_entry));
        } else {
            old_db_entry = self.second_db.write().replace(Arc::clone(&db_entry));
        }
```
Do you prefer this? If so, fine :)

Reply:
I'm not sure this code will survive once implemented in non-racy way in the first place.

Reply:
Here is another approach:
we can eliminate read/write locks altogether. The crux of the problem was the race between the solution finder and the old database removal. Both were trying to acquire the lock (read lock) for the current active database.

Solution: for the new database creation function, we can have a `static bool` that will alternate between true/false in each call. And depending on the value of that switch, we will be able to detect the old database, without reading the internals of it. Hence, no lock will be required. 

The downside of that approach would be: we will be injecting an implicit logic that is not obvious to an outsider. Comments should help though :)

The current way (read/write locks) seemed like a more direct way of solving this. Hence, I went with it.

And lastly, I can't think of any realistic scenario that a race will happen in it. So, I will claim my it is `non-racy` already :D

Reply:
Whether code is racy or not is defined by its structure, not the calling conventions. We may not hit it in practice, but I'm very uncomfortable with leaving that possibility in the code.